components/director/internal/domain/application/resolver.go

```diff
@@ -501,22 +526,51 @@ func (r *Resolver) Auths(ctx context.Context, obj *graphql.Application) ([]*grap
 
 	var out []*graphql.AppSystemAuth
 	for _, sa := range sysAuths {
+		if shouldSkip := r.checkApplicationOneTimeTokenIsInvalid(sa.Value, r.oneTimeTokenCfg); shouldSkip {
+			log.C(ctx).Debug("skipping one-time token due to its expiration or usage")
+			continue
+		}
+
 		c, err := r.sysAuthConv.ToGraphQL(&sa)
 		if err != nil {
 			return nil, err
 		}
 
+		if sa.Value.OneTimeToken != nil && sa.Value.OneTimeToken.Type == tokens.ApplicationToken {
+			oneTimeTokenForApplication, err := r.oneTimeTokenConv.ToGraphQLForApplication(*sa.Value.OneTimeToken)
+			if err != nil {
+				return nil, errors.Wrap(err, "while converting one-time token to graphql")
+			}
+
+			c.(*graphql.AppSystemAuth).Auth.OneTimeToken = &oneTimeTokenForApplication
+		}
 		out = append(out, c.(*graphql.AppSystemAuth))
 	}
 
 	err = tx.Commit()
 	if err != nil {
 		return nil, err
 	}
-
 	return out, nil
 }
 
+func (r *Resolver) checkApplicationOneTimeTokenIsInvalid(auth *model.Auth, config onetimetoken.Config) bool {
+	if auth == nil || auth.OneTimeToken == nil || auth.OneTimeToken.Type != tokens.ApplicationToken {
+		return false
+	}
+
+	if auth.OneTimeToken.Used {
+		return true
+	}
+
+	isExpired := auth.OneTimeToken.CreatedAt.Add(config.ApplicationExpiration).Before(r.timeService.Now())
+	if isExpired {
+		return true
+	}
```

Comment:
There is the following portion of code in file `validation_hydrator.go`:
```
expirationTime, err := vh.getExpirationTimeForToken(systemAuth)
	if err != nil {
		log.C(ctx).Error(err)
		respondWithAuthSession(ctx, w, authSession)
		return
	}

	if systemAuth.Value.OneTimeToken.CreatedAt.Add(expirationTime).Before(vh.timeService.Now()) {
		log.C(ctx).Infof("One Time Token with validity %s for system auth with ID %q has expired", expirationTime.String(), systemAuth.ID)
		respondWithAuthSession(ctx, w, authSession)
		return
	}
```

The code for expiration can be extracted in common function